pkg/registry/core/persistentvolume/storage/storage_test.go

```diff
@@ -208,3 +211,105 @@ func TestShortNames(t *testing.T) {
+			fg:                true,
+			oldPvHasTimestamp: false,
+			phaseChanges:      true,
+			timestampExpected: true,
+		},
+		"featureEnabled,oldTimestampUnset,phaseUnchanged,timestampUnset": {
+			initialPv:         validNewPersistentVolume("foo"),
+			fg:                true,
+			oldPvHasTimestamp: false,
+			phaseChanges:      false,
+			timestampExpected: false,
+		},
+		"featureDisabled,oldTimestampUnset,phaseChanged,timestampUnset": {
+			initialPv:         validNewPersistentVolume("foo"),
+			fg:                false,
+			oldPvHasTimestamp: false,
+			phaseChanges:      true,
+			timestampExpected: false,
+		},
+		"featureDisabled,oldTimestampUnset,phaseUnchanged,timestampUnset": {
+			initialPv:         validNewPersistentVolume("foo"),
+			fg:                false,
+			oldPvHasTimestamp: false,
+			phaseChanges:      false,
+			timestampExpected: false,
+		},
+		"featureDisabled,oldTimestampSet,phaseChanged,timestampSet": {
+			initialPv:         validNewPersistentVolume("foo"),
+			fg:                false,
+			oldPvHasTimestamp: true,
+			phaseChanges:      true,
+			timestampExpected: true,
+		},
+		"featureDisabled,oldTimestampSet,phaseUnchanged,timestampSet": {
+			initialPv:         validNewPersistentVolume("foo"),
+			fg:                false,
+			oldPvHasTimestamp: true,
+			phaseChanges:      false,
+			timestampExpected: true,
+		},
+	}
+
+	storage, statusStorage, server := newStorage(t)
+	defer server.Terminate(t)
+	defer storage.Store.DestroyFunc()
+	ctx := genericapirequest.WithNamespace(genericapirequest.NewContext(), metav1.NamespaceNone)
+	key, _ := storage.KeyFunc(ctx, "foo")
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.PersistentVolumeLastPhaseTransitionTime, tc.fg)()
+			defer storage.Storage.Delete(ctx, key, tc.initialPv, nil, rest.ValidateAllObjectFunc, false, nil)
+
+			newPv := tc.initialPv.DeepCopy()
+
+			if tc.oldPvHasTimestamp {
+				now := metav1.Now()
+				tc.initialPv.Status.LastPhaseTransitionTime = &now
+			}
+
+			err := storage.Storage.Create(ctx, key, tc.initialPv, nil, 0, false)
+			if err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+
+			if tc.phaseChanges {
+				newPv.Status.Phase = api.VolumeBound
+			}
+
+			_, _, err = statusStorage.Update(ctx, newPv.Name, rest.DefaultUpdatedObjectInfo(newPv), rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc, false, &metav1.UpdateOptions{})
+			if err != nil {
+				t.Fatalf("Unexpected error: %v", err)
+			}
+			obj, err := storage.Get(ctx, "foo", &metav1.GetOptions{})
+			if err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+			pvOut := obj.(*api.PersistentVolume)
+
+			if tc.timestampExpected {
+				if pvOut.Status.LastPhaseTransitionTime == nil {
+					t.Fatalf("Expected PV %s to have a timestamp but got nil", pvOut.GetName())
+				}
+			} else {
+				if pvOut.Status.LastPhaseTransitionTime != nil {
```

Comment:
I think it would be better to validate the exact timestamp. That way we can test whether or not we keep what the user sets in each scenario.

Take a look at the CSR example for how to stub out the `now` method for unit testing.

Reply:
Ack, added.